### PR TITLE
Small buffer optimize tls negotiated ALPN

### PR DESCRIPTION
**Pull request by ThadHouse**
The TLS negotiated ALPN is often very small in size (probably h3), however this was always allocating, and this allocation was always very small and stays live for a long time. Add a small buffer optimization to remove this allocation in the normal case.

Part of #1300

### CODE REVIEW

File: src/inc/quic_tls.h [hunk 1]
```diff
@@ -30,6 +30,11 @@ typedef struct CXPLAT_TLS_SECRETS CXPLAT_TLS_SECRETS;
 #define TLS_EXTENSION_TYPE_QUIC_TRANSPORT_PARAMETERS_DRAFT          0xffa5  // Host Byte Order
 #define TLS_EXTENSION_TYPE_QUIC_TRANSPORT_PARAMETERS                0x0039  // Host Byte Order
 
+//
+// The small buffer size for an ALPN to avoid allocations in the general case.
+//
+#define TLS_SMALL_ALPN_BUFFER_SIZE  16
```
Comment (ThadHouse) [inline src/inc/quic_tls.h:36]: Should this be 8 or 16 bytes? 8 will still fit `h3` and `h3-29`. Do we imagine any ALPN's longer the 7 bytes to happen in the real world?

**Comment by ThadHouse** [comment]
This PR was mostly created to investigate the perf impact. It might not be something we take.

File: src/inc/quic_tls.h [hunk 2]
```diff
@@ -291,6 +296,8 @@ typedef struct CXPLAT_TLS_PROCESS_STATE {
     //
     uint8_t* Buffer;
 
+    uint8_t SmallAlpnBuffer[TLS_SMALL_ALPN_BUFFER_SIZE];
```
Comment (nibanks) [inline src/inc/quic_tls.h:299]: Please put a comment.

File: src/inc/quic_tls.h (hunk 1 above)
Comment (nibanks) [reply src/inc/quic_tls.h:36]: I'm fine with 16.

**Review by nibanks — APPROVED** [review]
LGTM. Just a few comments.

**Comment by ThadHouse** [comment]
I realized I should probably add a test for this, since no tests will actually hit the long alpn case. I'll do that.

File: src/test/lib/HandshakeTest.cpp [hunk 3]
```diff
@@ -2075,3 +2075,101 @@ QuicTestConnectClientCertificate(
...
+
+void
+QuicTestInvalidAlpnLengths(
+    void
+    )
+{
+    int Lengths[] = { 0, QUIC_MAX_ALPN_LENGTH + 1 };
+    char AlpnBuffer[QUIC_MAX_ALPN_LENGTH + 3]; // + 3 so it can always be 0 terminated
+    for (int Len = 0; Len < (int)ARRAYSIZE(Lengths); Len++)
+    {
```
Comment (nibanks) [inline src/test/lib/HandshakeTest.cpp:2087]: nit: formatting (below as well)
```suggestion
    for (int Len = 0; Len < (int)ARRAYSIZE(Lengths); Len++) {
```

**Pull request merged**